docassemble/AssemblyLine/al_document.py

```diff
@@ -234,43 +238,43 @@ def safe_value(
         )
         max_chars = max(self.overflow_trigger - len(overflow_message), 0)
 
-        # If there are at least 2 lines, we can ignore overflow trigger.
-        # each line will be at least input_width wide
+        # If we preserve newlines, we need to account for max_lines, not just max_chars
         if preserve_newlines and max_lines > 1:
             if isinstance(value, str):
+                max_lines = self.max_lines(
+                    input_width=input_width, overflow_message_length=0
+                )
+                max_chars = self.overflow_trigger
                 # Replace all new line characters with just \n. \r\n inserts two lines in a PDF
                 value = re.sub(r"[\r\n]+|\r+|\n+", r"\n", value).rstrip()
-                line = 1
-                retval = ""
-                paras = value.split("\n")
-                para = 0
-                while line <= max_lines and para < len(paras):
-                    # add the whole paragraph if less than width of input
-                    if len(paras[para]) <= input_width:
-                        retval += paras[para] + "\n"
-                        line += 1
-                        para += 1
-                    else:
-                        # Keep taking the first input_width characters until we hit max_lines
-                        # or we finish the paragraph
-                        while line <= max_lines and len(paras[para]):
-                            retval += paras[para][:input_width]
-                            paras[para] = paras[para][input_width:]
-                            line += 1
-                        if not len(paras[para]):
-                            para += 1
-                            retval += "\n"
-                # TODO: check logic here to only add overflow message when we exceed length
-                if len(paras) > para:
-                    return (
-                        retval.rstrip() + overflow_message
-                    )  # remove trailing newline before adding overflow message
-                else:
-                    return retval
+                # textwrap.wrap does all the hard work for us here
+                return " ".join(
+                    wrap(
+                        value,
+                        width=input_width,
+                        max_lines=max_lines,
+                        replace_whitespace=False,
+                        placeholder=overflow_message,
```

Comment:
Not sure how this would interact with newlines in `value`, which people are likely to add in text areas I think.https://docs.python.org/3/library/textwrap.html#textwrap.TextWrapper.replace_whitespace says that it'll "cause strange output", which makes me a little hesitant.

Reply:
It doesn't cause strange output in our usage, which ends up removing the line breaks after any get added.

We just use textwrap.wrap to figure out how many lines should fit, but we let the pdf renderer handle the actual soft line breaks in the output.

Reply:
We aren't removing all of them, we're just collapsing things where multiple appear in a row. Test overflow of 90, and input:
```
At least 10% of people 

in this target audience do not know what this term means.
 
Therefore, you must avoid this term, explain this term, or link to an explanation.
```
Results in 3 lines of text, when there should only be 2, which could cause things to get cut off in the PDF.